datafusion/expr/src/expr.rs

```diff
@@ -358,19 +359,57 @@ impl ScalarUDF {
     }
 }
 
-/// Returns the field of a [`arrow::array::ListArray`] or [`arrow::array::StructArray`] by key
+/// Key of `GetIndexedFieldKey`.
+/// This structure is needed to separate the responsibilities of the key for `DataType::List` and `DataType::Struct`.
+/// If we use index with `DataType::List`, then we use the `list_key` argument with `struct_key` equal to `None`.
+/// If we use index with `DataType::Struct`, then we use the `struct_key` argument with `list_key` equal to `None`.
+/// `list_key` can be any expression, unlike `struct_key` which can only be `ScalarValue::Utf8`.
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct GetIndexedFieldKey {
```

Comment:
I found this API to be awkward -- I think we can do something to make it better. I am working on a PR now...

Reply:
filed https://github.com/apache/arrow-datafusion/issues/7193 to track improving this API